user_sync/rules.py

```diff
@@ -152,14 +169,45 @@ def read_desired_user_groups(self, mappings, directory_connector):
             
             filtered_directory_user_by_user_key[user_key] = directory_user
             self.get_organization_info(OWNING_ORGANIZATION_NAME).add_desired_group_for(user_key, None)
+
+            # set up groups in hook scope; the target groups will be used whether or not there's customer hook code
+            self.after_mapping_hook_scope['source_groups'] = set()
+            self.after_mapping_hook_scope['target_groups'] = set()
             for group in directory_user['groups']:
+                self.after_mapping_hook_scope['source_groups'].add(group) # this is a directory group name
                 dashboard_groups = mappings.get(group)
                 if (dashboard_groups != None):
                     for dashboard_group in dashboard_groups:
-                        organization_info = self.get_organization_info(dashboard_group.organization_name)
-                        organization_info.add_desired_group_for(user_key, dashboard_group.group_name)
-    
-        self.logger.info('Total directory users after filtering: %d', len(filtered_directory_user_by_user_key))        
+                        self.after_mapping_hook_scope['target_groups'].add(self.make_dashboard_group_qualified_name(dashboard_group.group_name, dashboard_group.organization_name))
+
+            # only if there actually is hook code: set up rest of hook scope, invoke hook, update user attributes
+            if (options['after_mapping_hook'] is not None):
+                self.after_mapping_hook_scope['source_attributes'] = directory_user['source_attributes'].copy()
+
+                # [TODO morr 2017-02-26]: Is there a more robust way to assemble the target attributes?
```

Comment:
What exactly are you worried about?  That there will be `directory_user` attributes that aren't on the Adobe side?  Then just use a comprehension constructor
```
{k: directory_user[k] for k in ('username', 'email', 'country', 'firstname', 'lastname')}
```